docs/developer/graceful-shutdown.md

```diff
@@ -0,0 +1,500 @@
+
+    // Initialize structured logging
+    log := log.InitLogs()
+    log.Info("Starting service")
+    defer func() {
+        log.WithField("uptime", time.Since(startTime)).Info("Service stopped")
+    }()
+
+    // Load configuration
+    cfg, err := config.LoadOrGenerate(config.ConfigFile())
+    if err != nil {
+        log.Fatalf("reading configuration: %v", err)
+    }
+
+    // Initialize observability (tracing)
+    tracerShutdown := instrumentation.InitTracer(log, cfg, "service-name")
+    defer func() {
+        if err := tracerShutdown(ctx); err != nil {
+            log.Fatalf("failed to shut down tracer: %v", err)
+        }
+    }()
+
+    // Set up signal handling
+    ctx, cancel := signal.NotifyContext(ctx, os.Interrupt, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGQUIT)
+    defer cancel()
+
+    // Initialize and start services
+    // ... service-specific initialization ...
+
+    // Wait for shutdown signal or error
+    select {
+    case <-ctx.Done():
+        if ctx.Err() == context.Canceled {
+            log.Info("Shutdown signal received, initiating graceful shutdown")
+        } else {
+            log.Infof("Context done (%v), initiating graceful shutdown", ctx.Err())
+        }
+    case err := <-serverErrors:
+        log.Errorf("Server error received: %v", err)
+        log.Info("Initiating shutdown due to server error")
+        cancel()
+    }
+
+    // Coordinated shutdown with timeout
+    const shutdownTimeout = 30 * time.Second
+    log.Infof("Starting coordinated shutdown (timeout: %v)", shutdownTimeout)
+
+    shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), shutdownTimeout)
+    defer shutdownCancel()
+
+    // Service-specific shutdown logic here
+    // ...
+
+    select {
+    case <-shutdownComplete:
+        log.Info("Graceful shutdown completed")
+    case <-shutdownCtx.Done():
+        log.Warn("Shutdown timeout exceeded, forcing exit")
+    }
+}
+```
+
+## Signal Handling
+
+### Supported Signals
+
+All services handle these signals consistently:
+
+| Signal | Value | Purpose | Source |
+|--------|-------|---------|--------|
+| `SIGINT` | 2 | Interactive interrupt (Ctrl+C) | `os.Interrupt` |
+| `SIGHUP` | 1 | Hangup/reload | `syscall.SIGHUP` |
+| `SIGTERM` | 15 | Graceful termination | `syscall.SIGTERM` |
+| `SIGQUIT` | 3 | Quit with core dump | `syscall.SIGQUIT` |
+
+### Implementation Notes
+
+**Use `os.Interrupt` instead of `syscall.SIGINT`:**
+```go
+// ✅ Correct: Cross-platform compatible
+signal.NotifyContext(ctx, os.Interrupt, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGQUIT)
+
+// ❌ Incorrect: Unix-specific
+signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGQUIT)
+```
+
+**Why `context.Background()` for shutdown context:**
+```go
+// ✅ Correct: Fresh context for shutdown operations
+shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), shutdownTimeout)
+
+// ❌ Incorrect: Original context is already cancelled
```

Comment:
this depends on usage we have a properly implemented shutdown timeout with the parent ctx above